api/products.thrift

```diff
@@ -0,0 +1,78 @@
+// -------------------------------------------------------------------------
+//                     The CodeChecker Infrastructure
+//   This file is distributed under the University of Illinois Open Source
+//   License. See LICENSE.TXT for details.
+// -------------------------------------------------------------------------
+
+include "shared.thrift"
+
+namespace py ProductManagement
+namespace js codeCheckerProductManagement
+
+
+/*
+struct PrivilegeRecord {
+  1: string   name,
+  2: bool     isGroup
+}
+typedef list<PrivilegeRecord> PrivilegeRecords
+*/
+
+struct DatabaseConnection {
+  1:          string engine,         // The database engine, such as "sqlite" or "postgresql".
+  2:          string host,
+  3:          i32    port,
+  4:          string username_b64,
+  5: optional string password_b64,   // Database password is NOT sent server->client!
+  6:          string database
+}
+
+/* ProductConfiguration carries administrative data regarding product settings. */
+struct ProductConfiguration {
+  1:          i64                id,
+  2:          string             endpoint,
+  3:          string             displayedName_b64,
+  4:          string             description_b64,
+  5: optional DatabaseConnection connection
+}
+typedef list<ProductConfiguration> ProductConfigurations
+
+/* Product carries data to the end user's product list and tasks. */
+struct Product {
```

Comment:
This struct is used (akin to `RunData`) to query the information for listing products (and checking their accessibility). Separating this would result in the client (both the GUI and the CLI) having to send multiple API requests to figure out if a product exists.

(In this case, from the client's perspective, `exists` means: `the product endpoint is registered in the configuration` (with some extra fluff such as displayed name and ID) + `the server can connect to the database` + `the user has rights to access the product`. If any of this doesn't hold, the clients will report an appropriate error, as no runs can be accessed in the product.)

Perhaps the next patch will remove `ąccessible`. It is not fleshed out yet, only a placeholder exists so the GUI can properly present a 🔒 symbol. Currently, the API handler for `getProducts()` hardcodes `accessible=True`.